### PR TITLE
GraphQL Subscriptions

### PR DESCRIPTION
**Pull request by dushaniw**
fix https://github.com/wso2/product-apim/issues/11701

### CODE REVIEW

File: modules/distribution/resources/api_templates/endpoint_template.xml [hunk 1]
```diff
@@ -128,17 +131,31 @@ optimize="$advance_ep.get("optimize")"
...
-        #http_endpoint( $ep_key $endpoints )
-        #elseif( $endpointClass == "ws")
-        #websocket_endpoint( $ep_key )
-        #elseif ( $endpointClass == "address" )
-        #address_endpoint($ep_key $endpoints)
-        #elseif ( $endpointClass == "failover" )
-        #set( $failover_endpoints = $endpoint_config.get("${type}_failovers"))
+
+        ##If graphql api
+        #if( $endpointClass == "graphql" )
```
Comment (dushaniw) [inline modules/distribution/resources/api_templates/endpoint_template.xml:136]: Code modification related to GQL Subscriptions
`
        ##If graphql api
        #if( $endpointClass == "graphql" )
            #if( $websocketResourceKey )
                ##If websocketResourceKey present in context then this is the graphQL subscription endpoint
                #set( $endpoint_config = $endpoint_config.get("ws") )
                #set( $isGQLSubscription = "true" )
            #else
                ##Else GraphQL query mutation http endpoint
                #set( $endpoint_config = $endpoint_config.get("http") )
            #end
            #set( $endpointClass = $endpoint_config.get("endpoint_type") )
        #end`

File: modules/distribution/resources/api_templates/velocity_template.xml [hunk 2]
```diff
@@ -167,99 +172,122 @@
...
-        #else
+                    #set( $topic= "$util.escapeXml($resource.getUriTemplate())" )
+                #else
         url-mapping="$resource.getUriTemplate()"
-        #set( $topic= "$resource.getUriTemplate()")
-        #end
+                    #set( $topic= "$resource.getUriTemplate()" )
+                #end
+
+                #if( $isSubscriptionAvailable )
```
Comment (dushaniw) [inline modules/distribution/resources/api_templates/velocity_template.xml:191]: Code Modification related to GQL Subscriptions.
```
                #if( $isSubscriptionAvailable )
                    #if( $resource.getMethodsAsString() == "" )
                        #set( $isGQLWSResource = true )
        binds-to="WebSocketInboundEndpoint, SecureWebSocketEP"
                    #else
        binds-to="default"
                         #set( $isGQLWSResource = false )
                    #end
                #end
```

File: modules/distribution/resources/api_templates/velocity_template.xml [hunk 3]
```diff
@@ -167,99 +172,122 @@
...
-#set( $filterRegex = "SANDBOX" )
-#else
-#set( $filterRegex = "PRODUCTION" )
-#end
-## AWS Lambda: start
-#if($endpoint_config.get("endpoint_type") == 'awslambda')
-#set( $accessKey = $!{endpoint_config.get("amznAccessKey")} )
-#set( $secretKey = $!{endpoint_config.get("amznSecretKey")} )
-#set( $resourceName = $!{resource.getAmznResourceName()} )
+                ##If graphql api
```
Comment (dushaniw) [inline modules/distribution/resources/api_templates/velocity_template.xml:231]: Code Modification related to GQL Subscriptions
```
                ##If graphql api
                #if ( $endpoint_config.get("endpoint_type") == "graphql" )
                    #if( $isGQLWSResource )
                        #set( $endpoint_config = $endpoint_config.get("ws") )
                    #else
                        #set( $endpoint_config = $endpoint_config.get("http") )
                    #end
                #end
```

File: modules/distribution/resources/api_templates/endpoint_template.xml [hunk 4]
```diff
@@ -103,7 +103,10 @@
 #timeout( $ep.get('config') )
 </http>
 <property name="ENDPOINT_ADDRESS" value="$util.escapeXml($endpointUrl)"/>
-        </endpoint>
+#if( $isGQLSubscription )
```
Comment (dushaniw) [inline modules/distribution/resources/api_templates/endpoint_template.xml:106]: Code Modification related to GQL Subscriptions
```
#if( $isGQLSubscription )
```

File: modules/distribution/resources/api_templates/velocity_template.xml [hunk 5]
```diff
@@ -142,11 +142,16 @@
      version="$!apiVersion"
      transports="$!transport"
      version-type="context"
-        #if($apiType == 'WS')
+        #if( $apiType == 'WS' )
         binds-to="WebSocketInboundEndpoint, SecureWebSocketEP"
+        #elseif( $apiType == 'GRAPHQL' )
```
Comment (dushaniw) [inline modules/distribution/resources/api_templates/velocity_template.xml:147]: Code Modification related to GQL Subscriptions
```
        #elseif( $apiType == 'GRAPHQL' )
            #if( $isSubscriptionAvailable )
        binds-to="WebSocketInboundEndpoint, SecureWebSocketEP, default"
            #end
```

File: modules/distribution/resources/api_templates/velocity_template.xml [hunk 6]
```diff
@@ -167,99 +172,122 @@
 </inSequence>
         </resource>
         #else
-        ## api not blocked
-        #set ($resourceNo = 0)
-        #foreach($resource in $resources)
+            ## api not blocked
```
Comment (dushaniw) [inline modules/distribution/resources/api_templates/velocity_template.xml:175]: Code Modification related to GQL Subscriptions
```
            ## api not blocked
            #set ( $resourceNo = 0 )
            #foreach( $resource in $resources )
                #set( $isGQLWSResource = false )
                #set( $endpoint_config = $originalEndPointConfig )
```

**Review by 1akshitha — APPROVED** [review]

**Review by tharindu1st — APPROVED** [review]

**Pull request merged**